feat:  implement opentelemetry for distributed tracing

namkyu1999 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you -->

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
-   [ ] Fixes #<issue number>
-   [ ] PR messages has document related information
-   [ ] Labelled this PR & related issue with `breaking-changes` tag
-   [ ] PR messages has breaking changes related information
-   [ ] Labelled this PR & related issue with `requires-upgrade` tag
-   [ ] PR messages has upgrade related information
-   [ ] Commit has unit tests
-   [ ] Commit has integration tests
-   [ ] E2E run Required for the changes

xavierW commented:
 邮件已收到，我会尽快阅读并回复，谢谢

smitthakkar96 commented on bin/experiment/experiment.go:
```diff
@@ -75,8 +77,22 @@ func init() {
 func main() {
+	ctx := context.Background()
+	// Set up Observability.
+	shutdown, err := telemetry.InitOTelSDK(ctx, true)
```
We might not have otel collector running during local development plus if our users when upgrading litmus might also not have OTEL setup. Let's keep this optional, if initialising otel sdk fails we can log this was warning wdyt?

namkyu1999 replied:
done

smitthakkar96 reviewed (commented):
This is a great start however I don't see any place where we are marking span as error in case of failure, do we plan to do this in future iterations? 

Based on the experiment context it might be good to explore what span attributes might be relevant. 

Lastly to be able to correlate logs with traces can we add traceID to logs?

namkyu1999 commented:
> This is a great start however I don't see any place where we are marking span as error in case of failure, do we plan to do this in future iterations?

I will record errors like below
```go
if err != nil {
	span.SetStatus(codes.Error, "operationThatCouldFail failed")
	span.RecordError(err)
}
```
But we have over 50 chaos faults, so it can be merged in the next PR (it is a time-consuming job)
> Based on the experiment context it might be good to explore what span attributes might be relevant.

Does this mean adding a span attribute to help with future queries by attribute?
> Lastly to be able to correlate logs with traces can we add traceID to logs?

sure, I will add it

smitthakkar96 commented:
> Does this mean adding a span attribute to help with future queries by attribute?

Span Attributes can be useful for queries but also they can serve as additional context when reviewing the spans. Let's take an example, you get an http request, you might wanna log the relevant headers, the request body and the path similarly in context of the experiments it might be helpful to log the parameters for the experiment, maybe the targets and any other data you think that might be helpful when looking at an error trace :)

namkyu1999 commented:
Thanks for your detailed comments @smitthakkar96 :) I am going to pave the foundation for collaborating with other contributors. Adopting a `contrib package` is also good idea

namkyu1999 commented:
I updated `contribute/developer-guide` folder for using `context.Context`

`context.Context` makes context propagation automatically so i fixed all of the functions to add `context.Context` as a parameter.

Here is a screen shot of distributed tracing. It's a result of `pod-delete` chaos fault.

<img width="1512" alt="Screenshot 2024-09-16 at 4 53 34 PM" src="https://github.com/user-attachments/assets/89b6d845-327e-4495-8547-d61a93d83189">

ispeakc0de commented on chaoslib/litmus/aws-ssm-chaos/lib/ssm/aws-ssm-chaos-by-tag.go:
```diff
@@ -10,16 +11,20 @@ import (
 // PrepareAWSSSMChaosByTag contains the prepration and injection steps for the experiment
-func PrepareAWSSSMChaosByTag(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+func PrepareAWSSSMChaosByTag(ctx context.Context, experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+	ctx, span := otel.Tracer(telemetry.TracerName).Start(ctx, "InjectAWSSSMChaosByTagChaos")
```
Can we use Fault suffix instead of Chaos? i.,e InjectAWSSSMChaosByTagFault

ispeakc0de commented on chaoslib/litmus/container-kill/helper/container-kill.go:
```diff
@@ -4,13 +4,14 @@ import (
+	"os/exec"
+	"strconv"
+	"time"
+
```
We have set OTEL_EXPORTER_OTLP_ENDPOINT env in helper but we haven't added any code changes in helper to use it

ispeakc0de commented on chaoslib/litmus/gcp-vm-disk-loss-by-label/lib/gcp-vm-disk-loss-by-label.go:
```diff
@@ -27,7 +30,9 @@ var (
 // PrepareDiskVolumeLossByLabel contains the prepration and injection steps for the experiment
-func PrepareDiskVolumeLossByLabel(computeService *compute.Service, experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+func PrepareDiskVolumeLossByLabel(ctx context.Context, computeService *compute.Service, experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+	ctx, span := otel.Tracer(telemetry.TracerName).Start(ctx, "InjectDiskVolumeLossByLabelChaos")
```
InjectGCPDiskVolumeLossByLabelChaos

ispeakc0de commented on chaoslib/litmus/gcp-vm-disk-loss/lib/gcp-vm-disk-loss.go:
```diff
@@ -28,7 +31,9 @@ var (
 // PrepareDiskVolumeLoss contains the prepration and injection steps for the experiment
-func PrepareDiskVolumeLoss(computeService *compute.Service, experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+func PrepareDiskVolumeLoss(ctx context.Context, computeService *compute.Service, experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+	ctx, span := otel.Tracer(telemetry.TracerName).Start(ctx, "InjectVMDiskLossChaos")
```
add gcp prefix

ispeakc0de commented on chaoslib/litmus/gcp-vm-instance-stop-by-label/lib/gcp-vm-instance-stop-by-label.go:
```diff
@@ -9,22 +10,26 @@ import (
 // PrepareVMStopByLabel executes the experiment steps by injecting chaos into target VM instances
-func PrepareVMStopByLabel(computeService *compute.Service, experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+func PrepareVMStopByLabel(ctx context.Context, computeService *compute.Service, experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+	ctx, span := otel.Tracer(telemetry.TracerName).Start(ctx, "InjectVMInstanceStopByLabelChaos")
```
check at other places as well

ispeakc0de commented on chaoslib/litmus/pod-fio-stress/lib/pod-fio-stress.go:
```diff
@@ -25,15 +28,17 @@ import (
 // PrepareChaos contains the chaos preparation and injection steps
-func PrepareChaos(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+func PrepareChaos(ctx context.Context, experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+	ctx, span := otel.Tracer(telemetry.TracerName).Start(ctx, "InjectPidFIOStressChaos")
```
InjectPodFIOStressChaos

ispeakc0de commented on chaoslib/litmus/pod-memory-hog-exec/lib/pod-memory-hog-exec.go:
```diff
@@ -28,7 +31,9 @@ import (
 // PrepareMemoryExecStress contains the chaos preparation and injection steps
-func PrepareMemoryExecStress(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+func PrepareMemoryExecStress(ctx context.Context, experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+	ctx, span := otel.Tracer(telemetry.TracerName).Start(ctx, "InjectPodMemoryHogChaos")
```
InjectPodMemoryHogExecChaos

ispeakc0de commented on bin/experiment/experiment.go:
```diff
@@ -75,8 +80,24 @@ func init() {
+	clients := cli.ClientSets{}
 
-	clients := clients.ClientSets{}
+	ctx, span := otel.Tracer(telemetry.TracerName).Start(ctx, "ExecuteExperiment")
```
As we are creating one more ctx context, it will shadow the above context var created with the same name. 
Which can cause the defer function not to work. Can we use a different var name here to avoid var shadow?

namkyu1999 replied:
fixed here [fix: avoid shadowing](https://github.com/litmuschaos/litmus-go/pull/706/commits/2d476f8837b8bd1cfb88dee9694a5fd27692cede)

ispeakc0de commented on chaoslib/litmus/aws-ssm-chaos/lib/ssm/aws-ssm-chaos-by-id.go:
```diff
@@ -24,7 +27,9 @@ var (
 // PrepareAWSSSMChaosByID contains the prepration and injection steps for the experiment
-func PrepareAWSSSMChaosByID(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+func PrepareAWSSSMChaosByID(ctx context.Context, experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets, resultDetails *types.ResultDetails, eventsDetails *types.EventDetails, chaosDetails *types.ChaosDetails) error {
+	ctx, span := otel.Tracer(telemetry.TracerName).Start(ctx, "InjectAWSChaosByIDChaos")
```
We init otel SDK if otelExporterEndpoint != "". Here we are calling Tracer function without checking if the endpoint is set or not. Will it handle negative cases implicitly or should we call this conditionally?

namkyu1999 replied:
Basically, They do nothing when Otel SDK is not init.

here's a otel [docs](https://opentelemetry.io/docs/specs/otel/trace/api/#behavior-of-the-api-in-the-absence-of-an-installed-sdk)

and here's a reference [PR](https://github.com/kubegems/kubegems/pull/398)

ispeakc0de commented on bin/experiment/experiment.go:
```diff
@@ -75,8 +80,24 @@ func init() {
+	if otelExporterEndpoint := os.Getenv(telemetry.OTELExporterOTLPEndpoint); otelExporterEndpoint != "" {
+		shutdown, err := telemetry.InitOTelSDK(ctx, true, otelExporterEndpoint)
+		if err != nil {
+			return
```
we can log error here
